FLT/HaarMeasure/DistribHaarChar/Padic.lean

```diff
@@ -0,0 +1,106 @@
+  change volume (H : Set ℚ_[p]) = ‖(x : ℚ_[p])‖₊ * volume (K : Set ℚ_[p])
+  -- This is true because `H` is a `‖x‖₊⁻¹`-index subgroup of `K`.
+  have hHK : H ≤ K := by
+    simpa [H, K, -Submodule.smul_le_self_of_tower]
+      using (.span _ 1 : Submodule ℤ_[p] ℚ_[p]).smul_le_self_of_tower (x : ℤ_[p])
+  have : H.FiniteRelIndex K :=
+    PadicInt.smul_submoduleSpan_finiteRelIndex_submoduleSpan (p := p)
+      (mem_nonZeroDivisors_iff_ne_zero.1 x.2) {1}
+  have H_relindex_Z : (H.relindex K : ℝ≥0∞) = ‖(x : ℚ_[p])‖₊⁻¹ :=
+    congr(ENNReal.ofNNReal $(PadicInt.smul_submoduleSpan_relindex_submoduleSpan (p := p) x {1}))
+  rw [← index_mul_addHaar_addSubgroup_eq_addHaar_addSubgroup hHK, H_relindex_Z, ENNReal.coe_inv,
+    ENNReal.mul_inv_cancel_left]
+  · simp
+  · simp
+  · simp
+  · simpa [H, K, Padic.submoduleSpan_padicInt_one]
+      using measurableSet_closedBall.const_smul (x : ℚ_[p]ˣ)
+  · simpa [K, Padic.submoduleSpan_padicInt_one] using measurableSet_closedBall
+
+/-- The distributive Haar character of the action of `ℚ_[p]ˣ` on `ℚ_[p]` is the usual p-adic norm.
+
+This means that `volume (x • s) = ‖x‖ * volume s` for all `x : ℚ_[p]` and `s : Set ℚ_[p]`.
+See `Padic.volume_padic_smul` -/
+lemma distribHaarChar_padic (x : ℚ_[p]ˣ) : distribHaarChar ℚ_[p] x = ‖(x : ℚ_[p])‖₊ := by
+  -- Write the RHS as the application of a monoid hom `g`.
+  let g : ℚ_[p]ˣ →* ℝ≥0 := {
+    toFun := fun x => ‖(x : ℚ_[p])‖₊
+    map_one' := by simp
+    map_mul' := by simp
+  }
+  revert x
+  suffices distribHaarChar ℚ_[p] = g by simp [this, g]
+  -- By density of `ℤ_[p]⁰` inside `ℚ_[p]ˣ`, it's enough to check that `distribHaarChar ℚ_[p]` and
+  -- `g` agree on `ℤ_[p]⁰`.
+  refine MonoidHom.eq_of_eqOn_dense (PadicInt.closure_nonZeroDivisors_padicInt (p := p)) ?_
+  -- But this is what we proved in `distribHaarChar_padic_padicInt`.
+  simp
+  ext x
+  simp [g]
+  rw [distribHaarChar_padic_padicInt]
+  rfl
+
+@[simp]
+lemma Padic.volume_padic_smul (x : ℚ_[p]) (s : Set ℚ_[p]) : volume (x • s) = ‖x‖₊ * volume s := by
+  obtain rfl | hx := eq_or_ne x 0
+  · simp [(finite_zero.subset s.zero_smul_set_subset).measure_zero]
+  · lift x to ℚ_[p]ˣ using hx.isUnit
+    rw [← distribHaarChar_padic, distribHaarChar_mul, Units.smul_def]
+
+@[simp] lemma Padic.volume_padicInt_smul (x : ℤ_[p]) (s : Set ℚ_[p]) :
+    volume (x • s) = ‖x‖₊ * volume s := by simpa [-volume_padic_smul] using volume_padic_smul x s
+
+@[simp] lemma PadicInt.volume_padicInt_smul (x : ℤ_[p]) (s : Set ℤ_[p]) :
+    volume (x • s) = ‖x‖₊ * volume s := by
+  simpa [-volume_padicInt_smul, ← image_coe_smul_set] using Padic.volume_padicInt_smul x ((↑) '' s)
+
+/-- The distributive Haar character of the action of `ℤ_[p]ˣ` on `ℤ_[p]` is the usual p-adic norm.
+
+This means that `volume (x • s) = ‖x‖ * volume s` for all `x : ℤ_[p]` and `s : Set ℤ_[p]`.
+See `PadicInt.volume_padicInt_smul` -/
+lemma distribHaarChar_padicInt (x : ℤ_[p]ˣ) : distribHaarChar ℤ_[p] x = ‖(x : ℤ_[p])‖₊ :=
```

Comment:
These are both equal to 1 so this lemma is very weird. Why not just prove it equals 1?